mockbroker_test.go

```diff
@@ -47,80 +72,134 @@ func (b *mockBroker) Addr() string {
+
+	go func() {
+		<-b.closing
+		safeClose(b.t, b.listener)
+	}()
+
+	wg := &sync.WaitGroup{}
+	for i := 0; ; i++ {
+		if conn, err = b.listener.Accept(); err != nil {
+			goto done
+		}
+		wg.Add(1)
+		go b.handleRequests(conn, i, wg)
 	}
-	reqHeader := make([]byte, 4)
+done:
+	wg.Wait()
+	Logger.Printf("*** mockbroker/%d: listener closed, err=%v", b.BrokerID(), err)
+}
+
+func (b *mockBroker) handleRequests(conn net.Conn, idx int, wg *sync.WaitGroup) {
+	defer wg.Done()
+	defer func() {
+		_ = conn.Close()
+	}()
+	Logger.Printf("*** mockbroker/%d/%d: connection opened", b.BrokerID(), idx)
+	var err error
+
+	abort := make(chan none)
+	defer close(abort)
+	go func() {
+		select {
+		case <-b.closing:
+			_ = conn.Close()
+		case <-abort:
+		}
+	}()
+
 	resHeader := make([]byte, 8)
-	for expectation := range b.expectations {
-		_, err = io.ReadFull(conn, reqHeader)
+	for {
+		req, err := decodeRequest(conn)
 		if err != nil {
-			return b.serverError(err, conn)
-		}
-		body := make([]byte, binary.BigEndian.Uint32(reqHeader))
-		if len(body) < 10 {
-			return b.serverError(errors.New("Kafka request too short."), conn)
-		}
-		if _, err = io.ReadFull(conn, body); err != nil {
-			return b.serverError(err, conn)
+			Logger.Printf("*** mockbroker/%d/%d: invalid request: err=%v, %v", b.brokerID, idx, err, spew.Sdump(req))
+			b.serverError(err)
+			goto done
 		}
 
 		if b.latency > 0 {
 			time.Sleep(b.latency)
 		}
 
-		response, err := encode(expectation)
+		res := b.requestHandler()(req)
+		if _, ok := req.body.(*FetchRequest); !ok {
+			Logger.Printf("*** mockbroker/%d/%d: served %+v -> %+v", b.brokerID, idx, req, res)
+		}
+		var encodedRes []byte
+		encodedRes, err = encode(res)
 		if err != nil {
-			return false
+			b.serverError(err)
+			goto done
 		}
-		if len(response) == 0 {
+		if len(encodedRes) == 0 {
 			continue
 		}
 
-		binary.BigEndian.PutUint32(resHeader, uint32(len(response)+4))
-		binary.BigEndian.PutUint32(resHeader[4:], binary.BigEndian.Uint32(body[4:]))
+		binary.BigEndian.PutUint32(resHeader, uint32(len(encodedRes)+4))
+		binary.BigEndian.PutUint32(resHeader[4:], uint32(req.correlationID))
 		if _, err = conn.Write(resHeader); err != nil {
-			return b.serverError(err, conn)
+			b.serverError(err)
+			goto done
 		}
-		if _, err = conn.Write(response); err != nil {
-			return b.serverError(err, conn)
+		if _, err = conn.Write(encodedRes); err != nil {
+			b.serverError(err)
+			goto done
 		}
 	}
-	if err = conn.Close(); err != nil {
-		return b.serverError(err, nil)
-	}
-	if err = b.listener.Close(); err != nil {
-		b.t.Error(err)
-		return false
-	}
-	return true
+done:
+	Logger.Printf("*** mockbroker/%d/%d: connection closed, err=%v", b.BrokerID(), idx, err)
 }
 
-func (b *mockBroker) serverError(err error, conn net.Conn) bool {
-	b.t.Error(err)
-	if conn != nil {
-		if err := conn.Close(); err != nil {
-			b.t.Error(err)
+func (b *mockBroker) requestHandler() func(req *request) (res encoder) {
+	b.handlerLock.Lock()
```

Comment:
I was not saying that this particular lock can resolve all racing conditions in tests, but it is necessary to make handler updates visible in the `MockBroker` goroutine.

I do not see how passing `map[string][]MockResponse` to `SetHandlerByMap` is going to help here. How would the mock broker decide when to stop using one `MockResponse` from the slice and start using another. Note that `MockResponse` is **not just an object** that should be returned, **but a builder** that constructs responses based on received requests. However in complex tests like this the response to be constructed depends not only on a received request but on the test stage too. So we can either write a complex `MockResponse` that is able to generate different responses depending on the test stage, or set different handlers depending on the test stage. I have chosen the later approach for it seemed to render more readable tests IMHO.

Note that there is still problem how to determine when one test stage completes and other starts. In most cases that can be done examining a PartitionConsumer Messages and Errors channels (although in this test it is mostly done via `time.Sleep`), but unfortunately some errors are not propagated to the Errors channel (e.g. ErrUnknownTopicOrPartition, ErrNotLeaderForPartition, ErrLeaderNotAvailable) and therefore there is no reliable way to determine the test stage boundary  and we are forced to use `time.Sleep`.

That said, it makes sense to leave the `SetHandlerByMap` signature and the respective lock as they are.